jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,465 @@
+import org.eclipse.jetty.server.HttpStream;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.AtomicBiInteger;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.util.NanoTime;
+import org.eclipse.jetty.util.ProcessorUtils;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.thread.Scheduler;
+import org.eclipse.jetty.util.thread.ThreadPool;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>A quality of service {@link Handler} that limits the number
+ * of concurrent requests, to provide good end-user experience in
+ * case descendant {@link Handler}s have limited capacity.</p>
+ * <p>This {@code Handler} limits the number of concurrent requests
+ * to the number configured via {@link #setMaxRequests(int)}.
+ * If more requests are received, they are suspended (that is, not
+ * forwarded to the child {@code Handler}) and stored in a priority
+ * queue.
+ * Priorities are determined via {@link #getPriority(Request)},
+ * that should return values between {@code 0} (the lowest priority)
+ * and positive numbers, typically in the range {@code 0-10}.</p>
+ * <p>When a request that is being processed completes, the suspended
+ * request that current has the highest priority is resumed.</p>
+ * <p>This {@link Handler} is ideal to avoid contending on slow/limited
+ * resources such as a JDBC connection pool, avoiding the situation
+ * where all server threads blocked contending on the limited
+ * resource, therefore leaving threads free to process other
+ * requests that do not require access to the limited resource.</p>
+ * <p>Requests are resumed in priority order, so that when the
+ * server is under load, and there are many requests suspended to
+ * be processed, high priority request are processed first.
+ * For example, load balancer "ping" requests may have the highest
+ * priority, followed by requests performed by admin users, etc.
+ * so that regardless of the load, "ping" and "admin" requests will
+ * always be able to access the web application.</p>
+ */
+@ManagedObject
+public class QoSHandler extends Handler.Wrapper
+{
+    private static final Logger LOG = LoggerFactory.getLogger(QoSHandler.class);
+    private static final String EXPIRED_ATTRIBUTE = QoSHandler.class.getName() + ".expired";
+
+    // hi=permits, lo=suspended.
+    private final AtomicBiInteger state = new AtomicBiInteger();
+    private final Forwarder forwarder = new Forwarder();
+    private final Map<Integer, Queue<Entry>> queues = new ConcurrentHashMap<>();
+    private final Set<Integer> priorities = new ConcurrentSkipListSet<>(Comparator.reverseOrder());
+    private CyclicTimeouts<Entry> timeouts;
+    private int maxRequests;
+    private Duration maxSuspend = Duration.ZERO;
+
+    public QoSHandler()
```

Comment:
Please add a constructor that takes a `Handler` to pass it to `super`. This helps chaining handlers.